src/interfaces/process.jl

```diff
@@ -191,15 +191,48 @@ Convenience function dispatching to [`incoming_particles`](@ref) or [`outgoing_p
     outgoing_particles(proc_def)
 
 """
-    number_particles(proc_def::AbstractProcessDefinition, ::ParticleDirection)
+    number_particles(proc_def::AbstractProcessDefinition, dir::ParticleDirection)
 
-Convenience function dispatching to [`number_incoming_particles`](@ref) or [`number_outgoing_particles`](@ref) depending on the given direction.
+Convenience function dispatching to [`number_incoming_particles`](@ref) or [`number_outgoing_particles`](@ref) depending on the given direction, returning the number of incoming or outgoing particles, respectively.
 """
 @inline number_particles(proc_def::AbstractProcessDefinition, ::Incoming) =
     number_incoming_particles(proc_def)
 @inline number_particles(proc_def::AbstractProcessDefinition, ::Outgoing) =
     number_outgoing_particles(proc_def)
 
+"""
+    number_particles(proc_def::AbstractProcessDefinition, dir::ParticleDirection, species::AbstractParticleType)
+
+Return the number of particles of the given direction and species in the given process definition.
+"""
+@inline function number_particles(
+    proc_def::AbstractProcessDefinition, dir::DIR, ::PT
+) where {DIR<:ParticleDirection,PT<:AbstractParticleType}
+    return count(x -> x isa PT, particles(proc_def, dir))
+end
+
+"""
+    number_particles(proc_def::AbstractProcessDefinition, particle::AbstractParticleStateful)
+    number_particles(proc_def::AbstractProcessDefinition, particle_type::Type{AbstractParticleStateful})
+
+Return the number of particles of the given particle's direction and species in the given process definition.
+"""
+@inline function number_particles(
+    proc_def::AbstractProcessDefinition, ::AbstractParticleStateful{DIR,PT,EL}
+) where {DIR<:ParticleDirection,PT<:AbstractParticleType,EL<:AbstractFourMomentum}
+    return number_particles(proc_def, DIR(), PT())
```

Comment:
It's not a good idea to call the direction and particle type parameter here. Better use the instances already instantiated in the particle-stateful using `particle_direction` and `particle_species`. This will also remove some noise from the function signature because you don't need to specify the type parameters in the first place.